Alerting: Disable /api/admin/pause-all-alerts with Unified Alerting

joeblubaugh opened the pull request:
/api/admin/pause-all-alerts only takes effect for legacy alerts. This
change returns a 403 if it's called when unified alerting is enabled.

Fixes #51729

joeblubaugh committed:
Alerting: Disable /api/admin/pause-all-alerts with Unified Alerting

/api/admin/pause-all-alerts only takes effect for legacy alerts. This
change returns a 403 if it's called when unified alerting is enabled.

Fixes #51729

papagian reviewed (commented):
LGTM but I have a question

joeblubaugh committed:
Switch from using the Unified Alerting setting to checking legacy

Checking the status of legacy alerting matches the semantics of the APIs
better.

joeblubaugh commented on pkg/api/alerting.go:
```diff
@@ -507,91 +507,107 @@ func (hs *HTTPServer) NotificationTest(c *models.ReqContext) response.Response {
-	if err := hs.SQLStore.GetAlertById(c.Req.Context(), &query); err != nil {
-		return response.Error(500, "Get Alert failed", err)
+func (hs *HTTPServer) PauseAlert(ngalertEnabled bool) func(c *models.ReqContext) response.Response {
+	if ngalertEnabled {
```
I've updated this to use `setting.AlertingEnabled`.

papagian commented on pkg/api/alerting.go:
```diff
@@ -507,91 +507,107 @@ func (hs *HTTPServer) NotificationTest(c *models.ReqContext) response.Response {
-	if err := hs.SQLStore.GetAlertById(c.Req.Context(), &query); err != nil {
-		return response.Error(500, "Get Alert failed", err)
+func (hs *HTTPServer) PauseAlert(legacyAlertingEnabled *bool) func(c *models.ReqContext) response.Response {
+	if legacyAlertingEnabled == nil || !*legacyAlertingEnabled {
```
Based on the following:
https://github.com/grafana/grafana/blob/62b0a8bae6abcbd75326d3310abebcf6a78eac78/pkg/setting/setting.go#L1427-L1431

https://github.com/grafana/grafana/blob/62b0a8bae6abcbd75326d3310abebcf6a78eac78/pkg/setting/setting_unified_alerting.go#L119-L122

I think that the only case `AlertingEnabled` is `nil` is when the `alerting.enabled` cannot be parsed and in that case the new alerting is enabled.
So, I think the check should become:
```suggestion
	if legacyAlertingEnabled != nil || !*legacyAlertingEnabled {
```

yuri-tceretian commented:
Although I am OK with this approach, I wonder whether it would be better to just not register the endpoint at all, i.e. update code here:
https://github.com/grafana/grafana/blob/1d2aa7c69bf7db8cd4a5b7514c83ab17966f1790/pkg/api/api.go#L461-L467

joeblubaugh commented:
@yuri-tceretian I prefer the more informative message in this case - we haven't deprecated the API in the past or provided any other advanced notice.

papagian commented:
> Although I am OK with this approach, I wonder whether it would be better to just not register the endpoint at all

I also like this suggestion unless we want to give better feedback to the users.
In any case I would suggest also to update [the docs](https://grafana.com/docs/grafana/latest/developers/http_api/admin/#pause-all-alerts) appropriately.

joeblubaugh committed:
Fix boolean pointer params.
